### PR TITLE
Make level ID required for level objects in React components

### PR DESCRIPTION
**Pull request by jmkulwik**
This is a sub-task of the progress data refactor.

Looking up level progress by level ID is an eventual goal of the progress data refactor. This is the first step. In the past, the level id would not be included in the level object. Now, that ID is a required part of the object and is verified through prop-types. Additionally, locations where a level is used now use levelType.

### Considerations
**Multi-page assessments**
These continue to work as expected with this change:
image
Example progression: http://localhost-studio.code.org:3000/s/csp5-2020/lockable/1/puzzle/2/page/1
https://studio.code.org/s/csp5-2020/lockable/1/puzzle/2/page/1 

**Sublevels**
These continue to work as expected with this change:
image
Example level: http://localhost-studio.code.org:3000/s/express-2020/stage/28/puzzle/2/sublevel/4
https://studio.code.org/s/express-2020/stage/28/puzzle/2/sublevel/4 

**Stage Extras**
Example Level: http://localhost-studio.code.org:3000/s/express-2020/stage/26/extras
And: http://localhost-studio.code.org:3000/s/express-2020/stage/28/extras
With these changes Stage Extras have feature parity with the current functionality in production including the following:
- The "header middle" component will always show the "stage extras" flag as green if the user is on the stage extras level and as gray if the user is not on the stage extras level.

## Links

- [investigation PR](https://github.com/code-dot-org/code-dot-org/pull/37930)
- [LP-1651](https://codedotorg.atlassian.net/browse/LP-1651)

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by jmkulwik** [comment]
I'm going to leave this one as-is and merge a follow-up with the sublevels work. Progress for bubble choice (sublevels) and rendering them are handled very differently and the PR for that will likely be similar in size to this one. Sublevels work correctly at this point in the process.

**Review by cforkish — CHANGES_REQUESTED** [review]
I'm noticing a few things that seem to be missing. Let me know if you determined any of the following to be unnecessary for this SOW.

1) progressRedux.currentLevelId is still being treated as a string, which is incompatible with level.id being a number.
2) currentLevelId is overloaded [here](https://github.com/code-dot-org/code-dot-org/blob/67b9417b49cb2b3472f5a7e1544b4da4a1850011/apps%2Fsrc%2Fcode-studio%2Fcomponents%2Fprogress%2FLessonProgress.jsx#L244) (from [here](https://github.com/code-dot-org/code-dot-org/blob/555ada5e41818c9c96cc776ddd462e81ace21c3e/dashboard%2Fapp%2Fviews%2Flayouts%2F_header.html.haml#L112)). search the investigation PR for "onLessonExtras" to see how i updated that to explicitly pass the boolean value from the server rather than passing it as an overloaded value for `currentLevelId`.
4) related to the above, should update `_header.html.haml` to pass back `#{script_level&.level_id || -1},` instead of `"#{uid}",` as current level id.
3) ids need to be added to summaries in `script_level.rb`

**Comment by jmkulwik** [comment]
@cforkish - Would love a second look when you get the chance!

File: apps/src/code-studio/progress.js [hunk 1]
```diff
@@ -78,6 +78,8 @@ progress.showDisabledBubblesAlert = function() {
  *   user, null otherwise
  * @param {boolean} stageExtrasEnabled Whether this user is in a section with
  *   stageExtras enabled for this script
+ * @param {boolean} isLessonExtras Boolean indicating we are not on a script
```
Comment (cforkish) [inline apps/src/code-studio/progress.js:81]: super minor NIT, but would be good to include your `currentPageNumber` annotation from above here and at `initializeStoreWithProgress`

File: dashboard/app/views/layouts/_header.html.haml [hunk 2]
```diff
@@ -105,14 +105,6 @@
     # don't trust outside content in parameter :puzzle_page - should be integer, so immediately call to_i
     puzzle_page = params[:puzzle_page] ? params[:puzzle_page].to_i : ApplicationHelper::PUZZLE_PAGE_NONE
 
-    if script_level
-      uid = params[:puzzle_page] ? "#{script_level.level_id}_#{puzzle_page - 1}" : script_level.level_id.to_s
-    else
-      # A bit of an abuse of uid, we dont have a specific level id in this case, so just use a hardcoded string
```
Comment (cforkish) [inline dashboard/app/views/layouts/_header.html.haml:111]: 🎉

**Review by cforkish — APPROVED** [review]
this all looks great! looks like you found some edge cases i never discovered in my big refactor, which will be good to know about in our H1 work. 👍 

now that you have such a thorough understanding of script levels and user progress really doesn't seem like the right time for you to leave LP..... :(

**Comment by jmkulwik** [comment]
Tests passed here. Only after this point are documentation changes: https://drone.cdn-code.org/code-dot-org/code-dot-org/2264

File: apps/src/code-studio/progress.js (hunk 1 above)
Comment (jmkulwik) [reply apps/src/code-studio/progress.js:81]: 👍

**Comment by jmkulwik** [comment]
@cforkish Reading through some comments on other PRs (especially @davidsbailey's comment here: https://github.com/code-dot-org/code-dot-org/pull/38403#issuecomment-755754936), I'd also recommend merging this one into staging on its own without waiting for the num -> string PR to be ready. This one's already big enough as is and is a good change to make on its own. It was originally intended to be merged completely separate from the num->string PR as they are two distinct changes. We've always tried to keep each PR small here and merged with its own comments. Merging the num->string PR into this one will be confusing to future code maintainers and isn't necessary for this change to work.

**Comment by jmkulwik** [comment]
For context, here's the doc that describes our general approach to PRs: https://docs.google.com/document/d/1-x-GcC2xWpUymU6YBWHzxmTAlASWQPMfAf8ASBqWlgA/edit

**Comment by jmkulwik** [comment]
And as a personal note, I'm not particularly comfortable with two large PRs with my name on them getting merged together into an even larger one before both go out to production. I'd much rather they go out separately on two separate days (or at least in two separate test builds) ensuring each works before the next is deployed.

**Comment by Hamms** [comment]
> And as a personal note, I'm not particularly comfortable with two large PRs with my name on them getting merged together into an even larger one before both go out to production

That does feel like a particularly brash way to tempt fate :upside_down_face:

**Pull request merged**